utilities/console.py

```
@@ -107,16 +115,40 @@ def disconnect(self):
 
-    def console_eof_sampler(self, func, command, timeout):
+    def _spawn_console(self) -> pexpect.fdpexpect.fdspawn:
+        """
+        Creates a pty pair and spawns virtctl via subprocess, returning an fdspawn
+        wrapping the master end.
+
+        Uses pty.openpty() + subprocess.Popen instead of pexpect.spawn to avoid
+        os.forkpty(), which is deprecated in multi-threaded processes (Python 3.12+).
+        """
+        master_fd, slave_fd = pty.openpty()
+        self._proc = subprocess.Popen(
```

Ruff (0.15.14)

[error] `subprocess` call: check for execution of untrusted input

(S603)